memcached.c

```diff
@@ -4191,6 +4287,837 @@ static inline void process_get_command(conn *c, token_t *tokens, size_t ntokens,
+        switch (opts[i]) {
+            case 'N':
+                of->locked = 1;
+                of->vivify = 1;
+                tokens++;
+                break;
+            case 'T':
+                tokens++;
+                of->locked = 1;
+                break;
+            case 'R':
+                of->locked = 1;
+                tokens++;
+                break;
+            case 'l':
+                of->la = 1;
+                of->locked = 1; // need locked to delay LRU bump
+                break;
+            case 'O':
+                tokens++;
+                break;
+            case 'k':
+                break;
+            case 'h':
+                of->locked = 1; // need locked to delay LRU bump
+                break;
+            case 'u':
+                of->no_update = 1;
+                break;
+            case 'q':
+                of->no_reply = 1;
+                break;
+            // mset-related.
+            case 'F':
+            case 'S':
+            case 'C':
+                tokens++;
+                break;
+            case 'I':
+                of->set_stale = 1;
+                break;
+        }
+    }
+
+    return tokens;
+}
+
+static void process_mget_command(conn *c, token_t *tokens, const size_t ntokens) {
+    char *key;
+    size_t nkey;
+    item *it;
+    char *opts;
+    char *fp = NULL;
+    char *p = c->wbuf;
+    size_t olen;
+    unsigned int i = 0;
+    int32_t rtokens = 0; // remaining tokens available.
+    struct _meta_flags of = {0}; // option bitflags.
+    uint32_t hv; // cached hash value for unlocking an item.
+    bool failed = false;
+    bool item_created = false;
+    bool won_token = false;
+    bool ttl_set = false;
+    char *errstr;
+
+    assert(c != NULL);
+
+    if (tokens[KEY_TOKEN].length > KEY_MAX_LENGTH) {
+        out_errstring(c, "CLIENT_ERROR bad command line format");
+        return;
+    }
+
+    key = tokens[KEY_TOKEN].value;
+    nkey = tokens[KEY_TOKEN].length;
+
+    // NOTE: final token has length == 0.
+    // KEY_TOKEN == 1. 0 is command.
+    rtokens = ntokens - 3; // cmd, key, final.
+
+    if (rtokens == 0) {
+        // Default flag options. Might not be the best idea.
+        opts = "sftv";
+        olen = 4;
+    } else {
+        // need to parse out the options.
+        opts = tokens[KEY_TOKEN + 1].value;
+        olen = tokens[KEY_TOKEN + 1].length;
+        rtokens--;
+    }
+
+    if (olen > MFLAG_MAX_OPT_LENGTH) {
+        out_errstring(c, "CLIENT_ERROR options flags are too long");
+        return;
+    }
+
+    // Copy opts into wbuf and advance pointer.
+    // We return the initial options + extra indicator flags.
+    // we reserve 4 bytes in front of the buffer, for up to three extra flags
+    // we can tag plus the initial space.
+    // This could be simpler by adding two iov's for the header line, but this
+    // is a hot path so trying to keep those to a minimum.
+    fp = c->wbuf + 6;
+    memcpy(fp, opts, olen);
```

Comment:
May be we can error if there are no valid options and move this after that check?

Reply:
ms is the only one currently enforcing that (needing at least 'S' flag) since mg and md can both operate with zero supplied flags. I'll think about it a bit.

Reply:
flag checking is now strict